### ISSUE TITLE
Trimming of video

### ISSUE THREAD

**Issue by kihaki** [linkedin/LiTr#43]
I have not found a way in the docs on how to trim the video down to a portion of the original (e.g. from x milliseconds to y milliseconds).

Seems to me that I could implement by creating a custom `MediaSource` and supplying it to the `transform` method. 

Am I on the right track here or is there an already built in way that I am currently missing?

**Comment by kihaki** [comment]
Right, so I have tried the following (modeled on the built in `MediaExtractorMediaSource`):

```
package de.app.transcoder

import android.content.Context
import android.media.MediaExtractor
import android.media.MediaFormat
import android.media.MediaMetadataRetriever
import android.net.Uri
import com.linkedin.android.litr.exception.MediaSourceException
import com.linkedin.android.litr.io.MediaSource
import com.linkedin.android.litr.utils.TranscoderUtils
import java.io.IOException
import java.nio.ByteBuffer

/**
 * An implementation of MediaSource, which wraps Android's [MediaExtractor] and allows for trimming of the input video
 */
class TrimmingMediaExtractorMediaSource(
    context: Context,
    uri: Uri,
    private val trimmingMs: LongRange
) : MediaSource {
    private val mediaExtractor: MediaExtractor = MediaExtractor()
    private val orientationHint: Int
    private val size: Long = TranscoderUtils.getSize(context, uri)

    private val MediaMetadataRetriever.orientation
        get() = extractMetadata(MediaMetadataRetriever.METADATA_KEY_VIDEO_ROTATION)
            ?.toInt() ?: 0

    init {
        val mediaMetadataRetriever = MediaMetadataRetriever()
        try {
            mediaExtractor.setDataSource(context, uri, null)
            mediaMetadataRetriever.setDataSource(context, uri)
        } catch (ex: IOException) {
            throw MediaSourceException(MediaSourceException.Error.DATA_SOURCE, uri, ex)
        }
        orientationHint = mediaMetadataRetriever.orientation
        mediaMetadataRetriever.release()

        // Set mediaExtractor to start at certain position
        mediaExtractor.seekTo(trimmingMs.first, MediaExtractor.SEEK_TO_PREVIOUS_SYNC)
    }

    override fun getOrientationHint(): Int = orientationHint

    override fun getTrackCount(): Int = mediaExtractor.trackCount

    override fun getTrackFormat(track: Int): MediaFormat = mediaExtractor.getTrackFormat(track)

    override fun selectTrack(track: Int) {
        mediaExtractor.selectTrack(track)
    }

    override fun seekTo(position: Long, mode: Int) {
        mediaExtractor.seekTo(position, mode)
    }

    override fun getSampleTrackIndex(): Int = mediaExtractor.sampleTrackIndex

    override fun readSampleData(buffer: ByteBuffer, offset: Int): Int {
        if (sampleTime > (trimmingMs.last * 1000L)) return -1 // End of trimmed range
        return mediaExtractor.readSampleData(buffer, offset)
    }

    override fun getSampleTime(): Long = mediaExtractor.sampleTime

    override fun getSampleFlags(): Int = mediaExtractor.sampleFlags

    override fun getSize(): Long = size

    override fun advance() {
        mediaExtractor.advance()
    }

    override fun release() {
        mediaExtractor.release()
    }
}
```

When in use, this leads to transcoding halting before reaching 100% and the resulting video file is not readable, which leads me to believe, that the end of file is not signaled correctly? Maybe some buffer flags are not set, since the extractor does not start from the beginning of the input file but from somewhere in between?

**Comment by izzytwosheds** [comment]
This is a great start! Thank you for taking initiative. 

What you are describing is quite possible, yes. I would probably start by adjusting a presentation time value in BufferInfo coming from MediaExtractor, to make is start at 0. 

Overall, trimming a video (and especially audio) might be more complex than creating a custom MediaSource. Trimming bounds might be between key frames for video and inside (not at the boundary) of an audio frame. This would require more complex logic, involving custom Decoder and possibly a custom Renderer. I will experiment on my end, too.

**Comment by grahamearley** [comment]
Hey @kihaki! I played around with your MediaSource and noticed that the transcode _does_ finish if you disable the audio track. This is because of what @izzytwosheds said about audio trimming being more difficult and requiring larger changes.

For now, I'm trying to get precise video trim working, before I look into audio trimming. When the transcode finished with your custom MediaSource, I noticed the video was still playing back its full length, untrimmed. Turns out this is because [`MediaExtractor.seekTo`](https://developer.android.com/reference/android/media/MediaExtractor#seekTo(long,%20int)) only seeks the selected tracks, and no tracks are selected yet at that point.

Moving the `seekTo` call into the `selectTrack` method (after `mediaExtractor.selectTrack(...)` is called) fixes this!

The start trim point is not precise here, since it will start at the nearest key frame before the specified trim start time. I'm now going to look at a way to solve that in the decoder.

**Comment by grahamearley** [comment]
Made some more progress! Instead of making a custom decoder, I ended up adding new logic to the `VideoTrackTranscoder`'s `resizeDecodedInputFrame()` method.

This is the logic that method currently uses to render frames (see the file [here](https://github.com/linkedin/LiTr/blob/35d161db1166215272084c31a2ea4e358bc79b8b/litr/src/main/java/com/linkedin/android/litr/transcoder/VideoTrackTranscoder.java#L189)):
```java
decoder.releaseOutputFrame(tag, true);
renderer.renderFrame(null, frame.bufferInfo.presentationTimeUs * MILLISECONDS_IN_SECOND);
```

This logic assumes we want to render every frame in the video (the first line's `true` parameter tells the decoder to render, and the second line performs the render). For trimming, there are frames we don't want to render! I updated the logic to this:

```java
long presentationTimeUs = frame.bufferInfo.presentationTimeUs;
boolean shouldRender = presentationTimeUs >= trimStartUs && presentationTimeUs <= trimEndUs;

decoder.releaseOutputFrame(tag, shouldRender);

if (shouldRender) {
    renderer.renderFrame(null, frame.bufferInfo.presentationTimeUs * MILLISECONDS_IN_SECOND);
}
```

And this gives us a precisely trimmed output!

I haven't cleaned anything up into more presentable code, but maybe next week I'll have some time to get a draft PR up. Right now, I'm hardcoding my trim start and end values in a few places, so I'll need to make a single spot for those parameters to be set.
